### PR TITLE
[trainer metrics] fix cpu mem metrics; reformat runtime metric

### PR DESCRIPTION
**Pull request by stas00**
This PR improves and fixes trainer metrics:

* reworks the general ram memory tracking replacing `tracemalloc`, with "sampling" via `psutil` - in particular for peak tracking using a thread. `tracemalloc` proved to not track anything but python memory allocations, so we were missing most of the general RAM from reports. Now we are reporting much more. (other than swapped out memory).
* adds important details to memory metrics docs
* moves `psutil` dependency from just-for-tests to the core. I tried to find a built-in python equivalent, but the closest that I found is `resource.getrusage(resource.RUSAGE_SELF).ru_maxrss` which doesn't report what we need and it's not cross-platform.
* reformats secs to be in `hh:mm:ss.msec` format so it's much easier to read the runtime metric

Discovered the `tracemalloc` limitation while tracking a huge memory leak in DeepSpeed when re-using deepspeed in the same process. My tests were consuming hundreds of MBs of general RAM and the metrics were reporting nothing.

before:
```
BS=4; PYTHONPATH=src USE_TF=0 python examples/seq2seq/run_translation.py --model_name_or_path \
t5-small --output_dir /tmp/zero3 --overwrite_output_dir --max_train_samples 64 --max_val_samples 64 \
--max_source_length 128 --max_target_length 128 --val_max_target_length 128 --do_train \
--num_train_epochs 1 --per_device_train_batch_size $BS --per_device_eval_batch_size $BS \
--learning_rate 3e-3 --warmup_steps 500 --predict_with_generate --logging_steps 0 --save_steps 0 \
--eval_steps 0 --group_by_length --adafactor --dataset_name wmt16 --dataset_config ro-en \
--source_lang en --target_lang ro --source_prefix "translate English to Romanian: "

***** train metrics *****
  epoch                      =    1.0
  init_mem_cpu_alloc_delta   =    3MB
  init_mem_cpu_peaked_delta  =    0MB
  init_mem_gpu_alloc_delta   =  230MB
  init_mem_gpu_peaked_delta  =    0MB
  train_mem_cpu_alloc_delta  =   60MB
  train_mem_cpu_peaked_delta =    0MB
  train_mem_gpu_alloc_delta  =  232MB
  train_mem_gpu_peaked_delta =  472MB
  train_runtime              = 5.5261
  train_samples              =     64
  train_samples_per_second   =  1.448
```

after this PR:

```
***** train metrics *****
  epoch                      =        1.0
  init_mem_cpu_alloc_delta   =     1298MB
  init_mem_cpu_peaked_delta  =      154MB
  init_mem_gpu_alloc_delta   =      230MB
  init_mem_gpu_peaked_delta  =        0MB
  train_mem_cpu_alloc_delta  =     3446MB
  train_mem_cpu_peaked_delta =        0MB
  train_mem_gpu_alloc_delta  =      232MB
  train_mem_gpu_peaked_delta =      472MB
  train_runtime              = 0:00:05.66
  train_samples              =         64
  train_samples_per_second   =      1.412
```

@sgugger, @LysandreJik

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -269,6 +269,7 @@ def run(self):
     deps["filelock"],  # filesystem locks, e.g., to prevent parallel downloads
     deps["numpy"],
     deps["packaging"],  # utilities from PyPA to e.g., compare versions
+    deps["psutil"],  # memory tracking
```
Comment (LysandreJik) [inline setup.py:272]: Usually not too keen on adding a new dependency to the core dependencies. Staying slim is a good way to prevent conflicts and to not break everything when there is an issue with another library; so we usually prefer adding soft dependencies to hard dependencies.

Is there something here that would prevent adding `psutil` as a soft dependency, rather than a hard one?

This case seems pretty narrow, as users that don't use the trainer (e.g., only use it for inference, or training with their custom loop/other libraries, only use pipelines, etc) don't need to have that in their system. 

This is how the hard dependencies are defined, with historical exceptions for regex and sacremoses, which could eventually be removed from the hard dependencies. Sacremoses is here because it's necessary to use for the XLM tokenizer, which, unfortunately, will never have a fast tokenizer.

File: src/transformers/trainer_utils.py [hunk 2]
```diff
@@ -282,8 +293,23 @@ class TrainerMemoryTracker:
     memory was needed to complete that stage.
 
     The reporting happens only for process of rank 0 and gpu 0 (if there is a gpu). Typically this is enough since the
-    main process does the bulk of work, but it could be not quite so if model parallel is used and then other gpus may
-    use a different amount of gpu RAM. Perhaps in the future this tracker will evolve to measure those too.
+    main process does the bulk of work, but it could be not quite so if model parallel is used and then other GPUs may
+    use a different amount of gpu memory. Perhaps in the future this tracker will evolve to measure those too.
```
Comment (LysandreJik) [inline src/transformers/trainer_utils.py:297]: Incredible docstring!

**Review by LysandreJik — APPROVED** [review]
This is a tremendous improvement relative to the previous implementation, great job @stas00!

The only thing I'm wondering about is the addition of `psutils` to the core dependencies. Having a hard requirement on it seems unnecessary. Would a soft dependency with a warning when not installed work for you? Happy to discuss if you think otherwise.

**Comment by LysandreJik** [comment]
Woops, didn't mean to approve

File: setup.py (hunk 1 above)
Comment (stas00) [reply setup.py:272]: Do you mean to add the check inside the metrics the first time it's used? Sure, that would work.

As I covered in OP I tried to find an alternative that is based on a built-in, but surprisingly didn't succeed.

**Review by sgugger — APPROVED** [review]
Thanks for the fix! Like @LysandreJik I would avoid adding `psutils` as a main dependency. We can have it come with a `is_psutils_available` and only compute the mems metrics when it's there.

**Comment by stas00** [comment]
> Thanks for the fix! Like @LysandreJik I would avoid adding `psutils` as a main dependency. We can have it come with a `is_psutils_available` and only compute the mems metrics when it's there.

Oh, I was thinking to assert to say to install it. They can disable the mem metrics flags if they don't want to install it.

So we have:

A. `assert("pip install psutil to use memory metrics")`
B. `return if not is_psutils_available()`


Either way works for me.

**Comment by sgugger** [comment]
I would use option B personally. Having an error because something is not installed is not something we like (cf wandb).

**Comment by stas00** [comment]
looks like I did something wrong with the runtime metrics - checking.
``` 
Trainer is attempting to log a value of "0:00:02.18" of type <class 'str'> for key "train/train_runtime" as a scalar. This invocation of Tensorboard's writer.add_scalar() is incorrect so we dropped this attribute.
```

**Comment by stas00** [comment]
@sgugger, so the only missing part is how can the user reach the extensive docs in `TrainerMemoryTracker` docstring? 

It's sort of an internal class. Perhaps I should move the docs elsewhere so that the user can understand what the memory metrics are? Perhaps I can do the following:

1. move the bulk of docstring explaning the metrics to `save_metrics` and then it'll automatically be in the right place
2. and add a note to `log_metrics` docstring to read the docstring of `save_metrics` for details
3. and add a note to `TrainerMemoryTracker` docstring to read the details in `save_metrics` for details

File: src/transformers/trainer_utils.py [hunk 3]
```diff
@@ -311,6 +331,18 @@ class TrainerMemoryTracker:
     }
 
     def __init__(self, skip_memory_metrics=False):
+
+        self.skip_memory_metrics = skip_memory_metrics
+
+        if not is_psutil_available():
+            # soft dependency on psutil
+            self.skip_memory_metrics = True
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:339]: Maybe add a logger.info telling the user they can `pip isntall psutil` to get automatic memory metrics?

**Comment by sgugger** [comment]
I think your approach for making the doc more visible is a good one, so I'm fine with it. Also add there that `pip install psutil` is necessary to get the memory metrics?

**Comment by stas00** [comment]
> I think your approach for making the doc more visible is a good one, so I'm fine with it. 

Great!

> Also add there that `pip install psutil` is necessary to get the memory metrics?

It's already there ;)

**Comment by sgugger** [comment]
Ah, missed it. Sorry about that!

File: src/transformers/trainer_utils.py (hunk 3 above)
Comment (stas00) [reply src/transformers/trainer_utils.py:339]: Since it will happen on every run, if they don't want it, that could be annoying.

It's already in `examples/_tests_requirements.txt` so at least users running the examples will already have it, the rest will be documented.

**Comment by stas00** [comment]
OK, docs moved/reshaped/cross linked from 2 places. Decided to put the main doc in `log_metrics` since that's where they are most "visual". 

If you could get one last look at the final version of the docs, that would be great. I expanded it a little bit more. I checked that they render well and cross-reference is a working link.

File: src/transformers/trainer_pt_utils.py [hunk 4]
```diff
@@ -675,6 +753,10 @@ def save_metrics(self, split, metrics, combined=True):
             The metrics returned from train/evaluate/predict
         combined (:obj:`bool`, `optional`, defaults to :obj:`True`):
             Creates combined metrics by updating ``all_results.json`` with metrics of this call
+
+    To understand the metrics please read the docstring of :obj:`transformers.Trainer.log_metrics()`. The only
```
Comment (sgugger) [inline src/transformers/trainer_pt_utils.py:757]:
```suggestion
    To understand the metrics please read the docstring of :meth:`~transformers.Trainer.log_metrics`. The only
```
This will make a link the user can click.

File: src/transformers/trainer_utils.py [hunk 5]
```diff
@@ -268,37 +271,9 @@ class TrainerMemoryTracker:
...
-    tracker will report incorrect info. If this `pytorch issue <https://github.com/pytorch/pytorch/issues/16266>`__
-    gets resolved it will be possible to change this class to be re-entrant. Until then we will only track the outer
-    level of ``train``, ``evaluate`` and ``predict`` methods. Which means that if ``eval`` is called during ``train``,
-    it's the latter that will account for its memory usage and that of the former.
-
-    This also means that if any other tool that is used along the :class:`~transformers.Trainer` calls
-    ``torch.cuda.reset_peak_memory_stats``, the gpu peak memory stats could be invalid. And the
-    :class:`~transformers.Trainer` will disrupt the normal behavior of any such tools that rely on calling
-    ``torch.cuda.reset_peak_memory_stats`` themselves.
+    To understand this class' intricacies please read the documentation of :obj:`transformers.Trainer.log_metrics()`.
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:276]:
```suggestion
    To understand this class' intricacies please read the documentation of :meth:`~transformers.Trainer.log_metrics`.
```

**Review by sgugger — APPROVED** [review]
Looks great!

File: src/transformers/trainer_pt_utils.py (hunk 4 above)
Comment (stas00) [reply src/transformers/trainer_pt_utils.py:757]: Actually :obj:`transformers.Trainer.log_metrics()` creates a clickable link too. 

But if the other approach is the one we use everywhere I will gladly switch to that. Thank you!

Comment (sgugger) [reply src/transformers/trainer_pt_utils.py]: Oh I didn't know obj tried to resolve the differences, that works for me then!

Comment (stas00) [reply src/transformers/trainer_pt_utils.py]: Let's keep things consistent across the doc, so your suggestion seems to be the one then.

I'm still wanting to resolve that cross-referencing issue for inherited methods so that they are clickable, but didn't find enough time to sort it out. But that's when I learned about `the :obj:transformers.Trainer.log_metrics()`  method.

**Review by LysandreJik — APPROVED** [review]
This looks great! Thanks for implementing that @stas00

**Comment by stas00** [comment]
Hmm, I'm having second thoughts about skipping and not asserting if `psutil` is unavailable. Since there is a function flag to skip memory metrics, if the flag is `False` and we skip the metrics, that's not super intuitive. So if a user doesn't want the memory metrics they don't have to install `pustil` but can simply disable the metrics by setting the skip flag to `True`.

Perhaps it'd be agreeable with you to change the behavior to option A. in https://github.com/huggingface/transformers/pull/10937#issuecomment-809529412, i.e. to assert.

**Comment by sgugger** [comment]
I'd personally really like to avoid the script failing even if you can set an argument to avoid that.
